rand_distr/src/skew_normal.rs

```diff
@@ -0,0 +1,218 @@
+// Copyright 2021 Developers of the Rand project.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// https://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or https://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! The Skew Normal distribution.
+
+use crate::{Distribution, StandardNormal};
+use core::fmt;
+use num_traits::Float;
+use rand::Rng;
+
+/// The [skew normal distribution] `SN(location, scale, shape)`.
+///
+/// The skew normal distribution is a generalization of the
+/// [`Normal`] distribution to allow for non-zero skewness.
+///
+/// It has the density function
+/// `f(x) = 2 / scale * phi((x - location) / scale) * Phi(alpha * (x - location) / scale)`
+/// where `phi` and `Phi` are the density and distribution of a standard normal variable.
+///
+/// # Example
+///
+/// ```
+/// use rand_distr::{SkewNormal, Distribution};
+///
+/// // location 2, scale 3, shape 1
+/// let skew_normal = SkewNormal::new(2.0, 3.0, 1.0).unwrap();
+/// let v = skew_normal.sample(&mut rand::thread_rng());
+/// println!("{} is from a SN(2, 3, 1) distribution", v)
+/// ```
+///
+/// # Implementation details
+///
+/// We are using the algorithm from [A Method to Simulate the Skew Normal Distribution].
+///
+/// [`skew normal distribution`]: https://en.wikipedia.org/wiki/Skew_normal_distribution
+/// [A Method to Simulate the Skew Normal Distribution]:
+/// 	Ghorbanzadeh, D. , Jaupi, L. and Durand, P. (2014)
+/// 	[A Method to Simulate the Skew Normal Distribution](https://dx.doi.org/10.4236/am.2014.513201).
+/// 	Applied Mathematics, 5, 2073-2076.
+#[derive(Clone, Copy, Debug)]
+#[cfg_attr(feature = "serde1", derive(serde::Serialize, serde::Deserialize))]
+pub struct SkewNormal<F>
+where
+    F: Float,
+    StandardNormal: Distribution<F>,
+{
+    location: F,
+    scale: F,
+    shape: F,
+}
+
+/// Error type returned from `SkewNormal::new`.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Error {
+    /// The scale parameter is not finite or not positive.
+    BadScale,
+    /// The shape parameter is not finite.
+    BadShape,
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(match self {
+            Error::BadScale => "scale parameter is non-finite in skew normal distribution",
+            Error::BadShape => "shape parameter is non-finite in skew normal distribution",
+        })
+    }
+}
+
+#[cfg(feature = "std")]
+#[cfg_attr(doc_cfg, doc(cfg(feature = "std")))]
+impl std::error::Error for Error {}
+
+impl<F> SkewNormal<F>
+where
+    F: Float,
+    StandardNormal: Distribution<F>,
+{
+    /// Construct, from location, scale and shape.
+    ///
+    /// Parameters:
+    ///
+    /// -   location (unrestricted)
```

Comment:
Can it be `nan`?

Reply:
Yes, it can! :) Just like in `Normal`

Reply:
I added a test for this. All samples will be NAN.